plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs

```diff
@@ -0,0 +1,159 @@
+that to split the map of parameters into two maps.
+
+-}
+
+-- | See point 3 of Note [Cost model parameters]
+-- Some of the numbers in the cost model are of the order or 1e-7 or 1e-8, and
+-- we need a large scale factor to avoid truncating a significant number of
+-- digits.
+scaleFactor :: S.Scientific
+scaleFactor = 1e20
+
+-- See Note [Cost model parameters]
+type CostModelParams = Map.Map Text.Text Integer
+
+-- See Note [Cost model parameters]
+-- | Extract the model parameters from a model.
+extractParams :: ToJSON a => a -> Maybe CostModelParams
+extractParams cm = case toJSON cm of
+    Object o ->
+        let
+            flattened = flattenObject "-" o
+            toScaledInteger :: S.Scientific -> Integer
+            toScaledInteger n = floor (n*scaleFactor)
+            scaledNumbers = HM.mapMaybe (\case { Number n -> Just $ toScaledInteger n; _ -> Nothing }) flattened
+            mapified = Map.fromList $ HM.toList scaledNumbers
+        in Just mapified
+    _ -> Nothing
+
+
+-- See Note [Cost model parameters]
+-- | Update a model by overwriting the parameters with the given ones.
+applyParams :: (FromJSON a, ToJSON a) => a -> CostModelParams -> Maybe a
+applyParams cm params = case toJSON cm of
+    Object o ->
+        let
+            hashmapified = HM.fromList $ Map.toList params
+            scaledNumbers = fmap (\n -> Number $ fromIntegral n / scaleFactor) hashmapified
+            flattened = flattenObject "-" o
+            -- this is where the overwriting happens, this is left-biased
+            merged = HM.union scaledNumbers flattened
+            unflattened = unflattenObject "-" merged
+        in case fromJSON (Object unflattened) of
+            Success a -> Just a
+            Error _   -> Nothing
+    _ -> Nothing
+
+-- | Parameters for a machine step model and a builtin evaluation model bundled together.
+data SplitCostModelParams =
+    SplitCostModelParams {
+      machineParams :: CostModelParams
+    , builtinParams :: CostModelParams
+    }
+
+-- | Split a CostModelParams object into two subobjects according to some prefix:
+-- see item 5 of Note [Cost model parameters].
+splitParams :: Text.Text -> CostModelParams -> SplitCostModelParams
+splitParams prefix params =
+    let machineparams = Map.filterWithKey (\k _ ->       Text.isPrefixOf prefix k) params
+        builtinparams = Map.filterWithKey (\k _ -> not $ Text.isPrefixOf prefix k) params
+    in SplitCostModelParams machineparams builtinparams
+
+-- | Given a CostModel, produce a single map containing the parameters from both components
+extractCostModelParams :: ToJSON machinecosts => CostModel machinecosts -> Maybe CostModelParams
+extractCostModelParams model =
+    case ( extractParams (machineCostModel model)
+         , extractParams (builtinCostModel model) )
+    of (Just machineparams, Just builtinparams) -> Just $ Map.union machineparams builtinparams
+       _                                        -> Nothing
```

Comment:
Why not the regular do-notation thing?

```haskell
do
    machineparams <- extractParams (machineCostModel model)
    builtinparams <- extractParams (builtinCostModel model)
    Just $ Map.union machineparams builtinparams
```

or even just

```haskell
Map.union <$> extractParams (machineCostModel model) <*> extractParams (builtinCostModel model)`
```

for that matter.